File: 23.07.2024/Connected Cells in a Grid.java
import java.util.Scanner;

public class ConnectedCells {
    static int dfs(int[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == 0) return 0;
        grid[i][j] = 0;
        int size = 1;
        for (int x = -1; x <= 1; x++) for (int y = -1; y <= 1; y++) size += dfs(grid, i + x, j + y);
        return size;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), m = sc.nextInt();
        int[][] grid = new int[n][m];
        for (int i = 0; i < n; i++) for (int j = 0; j < m; j++) grid[i][j] = sc.nextInt();
        int maxArea = 0;
        for (int i = 0; i < n; i++) for (int j = 0; j < m; j++) maxArea = Math.max(maxArea, dfs(grid, i, j));
        System.out.println(maxArea);
    }
}

File: 23.07.2024/Connected Cells in a Grid.py
def max_region(grid):
    def dfs(x, y):
        if x < 0 or y < 0 or x >= len(grid) or y >= len(grid[0]) or not grid[x][y]:
            return 0
        grid[x][y] = 0
        return 1 + sum(dfs(x+dx, y+dy) for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)])

    return max(dfs(x, y) for x in range(len(grid)) for y in range(len(grid[0])))

grid = [
    [1, 1, 0, 0],
    [0, 1, 1, 0],
    [0, 0, 1, 0],
    [1, 0, 0, 0]
]
print(max_region(grid))
